### DISCUSSION TITLE
Generics type inference?

### DISCUSSION THREAD

**Discussion by p0lunin** [tact-lang/spec#11] [General]
In most languages generics are implemented using special meta-language with special syntax with special brackets such as `<>` or `[]`. In Tact generics will implemented using compile-time functions like other things. In our case there are ergonomic problems with type inference. Suppose we have following parametrized function:
```
let func(T: Type) = fn(value: T) { ... };
// Will desugared to
let func = fn func(T: Type) -> Function { fn(value: T) { ... } };
```

When this declaration is desugared we have function that returns function. So, if we want to call `func` and get output value we need to call it twice - first time we must pass `T` which is generic type, second time we must pass all the function arguments:
```
let test = func(Type)(value);
```

While other languages provide type inference in this case because they can infer `T` type we cannot use this approach in such way because we have simple functions that we cannot avoid directly - if we call `func` only once it will be parsed as if we passing `value` argument where function expects to take argument with `Type` type.

It is also applied to the method calls. Suppose we have method `fn new() { ... }` which is implemented on the `Option(A)` type with generic `A`. How its call should looks like? If we simple call `Option::new()` then compiler will see that `Option` is a function of type `Type -> Type` and it will not accept this call, because `Type -> Type` type does not have method `new`.

So, we need to answer on the question: Do we need type inference for generics? If yes, how it should looks like?

**Comment by oleganza** [comment]
Seems like the usual "inference" feature as seen in Haskell, ML, Rust, Swift is based on the constraint system language, which we deliberately reject in our design. I mean that in those languages you have a constraint system and for each value you can solve the constraint equations and fill in the type.

**Comment by oleganza** [comment]
Small nitpick regarding your example of `Option::new()`: that's a good example on its own, but it's not the method in the usual sense. 

If you have an instance of this type, it will have concretely instantiated type `Option<T>` where the "true" method `unwrap` is implemented concretely to return `T`, so no explicit type declaration is necessary in such case: `let x = maybe_x.unwrap();`

**Comment by oleganza** [comment]
Maybe we won't have this problem if we erase distinction between meta-types, types and values using set semantics. 

Let's say we stipulate:
* What we call `Type` now is a set `Any` which includes all possible values, 
* `Int` is a subset of set `Any` that includes integer values,
* `1` is a one-item subset of set `Int` that includes one integer value.
* `Never` is an empty set Ø.

Function calls on values possible when the set is reduced to one item. Otherwise, those function calls are partially-applied:

```
fn add(a: Int257, b: Int257) { 
  a + b
}

add(Int8, 1) => fn(a: Int8) { a + 1 }
```

In this model we won't have an awkward wrapper function for `func(T)(...)` — we'd just pass concrete values that match the type. In case of `Option(T:Any)` we would simply hold "any value". And compiler would have to specialize code for various kinds of concrete primitive values.

However, i doubt that would work because what "type" should `Option(1)` have? In the sense that it should be compatible with `Option(2)` and `Option(2^32)`. We may find ourselves inventing constraint systems going this path.

**Reply by cmxog** [reply]
I am still to process this thread, but I wanted to note that I did have a vague plan for introducing some kind of `Any` type already. I've had some form of its roots with `HoleType` (intended for inference originally) https://github.com/tact-lang/tact/blob/main/lib/lang.ml#L65

**Comment by cmxog** [comment]
What might be useful in this specific case (parameterized functions) is to explore where we will use them most often and see if explicit passage of types as shown is going to be a significant burden.

I did run into this kind of problem when working on [newtype enablement pre-proposal](https://github.com/tact-lang/spec/discussions/6#discussioncomment-2718042) and it looks like what worked much better for most cases where I consdered a parameterized function, was either a parameterized type or a parameterized implementation of the interface. Of course, that is still limited to a known set, and at some point, an indication of a type is still required in absence of any type inference we can have there.

While I don't have a solution for this type of type inference at this moment, we can at least consider the following. There's this `TypeOf` function, so you can potentially rewrite the code you've given in the example to:

```
let test = func(TypeOf(value))(value);
```

which, by itself, is not a significant improvement, but it could lead to a shorthand form, perhaps; or a deduction rule that would allow to simplify the syntax? I am spitballing here, and I don't want to introduce new punctuation, but for the sake of the argument, what if the above code could be produced with something like this?

```
let test = func(_)(value);
```

Other ideas are welcome. This is just something that came to my mind that doesn't look *too* out of place.

**Reply by p0lunin** [reply]
`TypeOf(value)` is more closer to C++s `decltype` because `value` can be any expression such as `1 + 2` or `fcall(args)`. I do not think this is a good choice to solve such problems. Suppose you have function call:
```
fcall(1 + 2, do_stuff(variable), fn foo() { ... });
```
So, to call it using `TypeOf` function we should write something like:
```
fcall(TypeOf(1 + 2), TypeOf(do_stuff(variable)), TypeOf(fn foo() { ... }))(1 + 2, do_stuff(variable), fn foo() { ... });
```
Which is not an ergonomic way to solve such problem.

`_` syntax looks better but in this case when we have many generics it also do not looks ergonomic:
```
fcall(_, _, _)(1 + 2, do_stuff(variable), fn foo() { ... });
```
This `fcall(_, _, _)` looks a bit strange.
